brave/brave-core
# Use BalanceScanner contracts for ETH asset discovery

Resolves https://github.com/brave/brave-browser/issues/28203
Resolves https://github.com/brave/brave-browser/issues/27666

Note: The base branch is `get-erc20-token-balances`, not `master`.

Use the JsonRpcService::GetERC20TokenBalances API added in https://github.com/brave/brave-core/pull/16955 instead of the current approach that uses eth_getLogs to discover assets.

After the changes, one request will be made per address per chain to discover ETH assets, instead of one per chain. This reflects how it is done for Solana asset discovery.  Previously DiscoverEthAssets took a chain ID param, and made just one request for that chain.  Now DiscoverEthAssets just takes the list of addresses, and makes one request per chain, using a base::BarrierCallback to merge the results.  Consequently, CompleteDiscoverAssets is now called once per CoinType, rather than once per chain ID. 

Currently, the code makes no attempt to make these requests sequentially in batches, or spread the requests out over time.

## Submitter Checklist:

  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`

## Reviewer Checklist:


## After-merge Checklist:

  changes has landed on

## Test Plan:
1. Add an Ethereum account that has an asset in our registry.  The asset should automatically appear in the user asset list.  Repeat for a different EVM network.
2. Send an asset in our registry to an Ethereum account that does not already hold that asset.  Refresh the portfolio page (might need to wait a minute before doing this in order for the the rate limiting to expire if necessary).  The asset should appear in the user asset list.
